src/Bicep.LangServer/Snippets/SnippetCache.cs

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Immutable;
+using System.IO;
+using System.Text.Json;
+using Bicep.Core.Resources;
+
+namespace Bicep.LanguageServer.Snippets;
+
+public record SnippetCache(
+    ImmutableDictionary<ResourceTypeReference, SnippetCache.SnippetCacheSnippet> ResourceTypeReferenceInfoMap,
+    ImmutableDictionary<ResourceTypeReference, string> ResourceTypeReferenceToDependentsMap,
+    ImmutableDictionary<ResourceTypeReference, ImmutableArray<ResourceTypeReference>> ResourceTypeReferenceToChildTypeSymbolsMap,
+    ImmutableArray<Snippet> TopLevelNamedDeclarationSnippets)
+{
+    private static JsonSerializerOptions GetSerializerOptions()
+    {
+        var options = new JsonSerializerOptions();
+        options.Converters.Add(new ResourceTypeReferenceJsonConverter());
+
+        return options;
+    }
+
+    public record SnippetCacheSnippet(
+        string Prefix,
+        string Identifier,
+        string BodyText,
+        string Description);
+
+    public static SnippetCache Deserialize(Stream stream)
+        => JsonSerializer.Deserialize<SnippetCache>(stream, GetSerializerOptions())
+            ?? throw new InvalidOperationException($"Failed to deserialize embedded snippet cache");
+
+    public static string Serialize(SnippetCache cache)
+        => JsonSerializer.Serialize<SnippetCache>(cache, GetSerializerOptions());
+
+    private static readonly Lazy<SnippetCache> lazySnippetCache = new(FromManifestInternal);
+    private static SnippetCache FromManifestInternal()
+    {
+        var assembly = typeof(SnippetCache).Assembly;
+        var stream = assembly.GetManifestResourceStream("Files/SnippetCache.json")
+            ?? throw new InvalidOperationException($"Failed to find embedded snippet cache");
```

Comment:
Not for this PR, but we have many `Assemebly.GetManifestResourceStream() ?? throw ...` in the codebase now, so it might be worth extracting a helper method. I've created https://github.com/Azure/bicep/issues/8975 to track this.